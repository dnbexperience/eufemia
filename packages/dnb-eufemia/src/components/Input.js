/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index input to autogenerate all the components and patterns
 * Used by "prepareInputs"
 */

import Input from './input/Input'
export * from './input/Input'
export default Input
