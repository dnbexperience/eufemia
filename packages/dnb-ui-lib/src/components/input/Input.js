/**
 * Web Input Component
 *
 */

import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { DefaultIconSize } from '../icon'
import Button from '../button/Button'
import FormLabel from '../form-label/FormLabel'
import FormStatus from '../form-status/FormStatus'
import {
  registerElement,
  validateDOMAttributes,
  processChildren,
  pickRenderProps,
  dispatchCustomElementEvent
} from '../../shared/component-helper'

const renderProps = {
  on_change: null,
  on_submit: null,
  on_focus: null,
  on_blur: null,
  on_submit_focus: null,
  on_submit_blur: null
}

export const propTypes = {
  type: PropTypes.string,
  size: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  id: PropTypes.string,
  label: PropTypes.string,
  status: PropTypes.string,
  status_state: PropTypes.string,
  autocomplete: PropTypes.oneOf(['on', 'off']),
  search_button_title: PropTypes.string,
  placeholder: PropTypes.string,
  description: PropTypes.string,
  align: PropTypes.string,
  extra_information: PropTypes.string,
  disabled: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  class: PropTypes.string,
  input_class: PropTypes.string,
  input_typo: PropTypes.string,
  attributes: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  // React props
  className: PropTypes.string,
  inputElement: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func
  ]),

  // Web Component props
  custom_element: PropTypes.object,
  custom_method: PropTypes.func,
  on_change: PropTypes.func,
  on_submit: PropTypes.func,
  on_focus: PropTypes.func,
  on_blur: PropTypes.func,
  on_submit_focus: PropTypes.func,
  on_submit_blur: PropTypes.func
}

export const defaultProps = {
  type: 'text',
  size: null,
  value: null,
  id: null,
  label: null,
  status: null,
  status_state: 'error',
  autocomplete: 'off',
  search_button_title: '',
  placeholder: null,
  description: null,
  align: null,
  extra_information: null,
  disabled: false,
  input_class: null,
  input_typo: 'dnb-typo-number--lining',
  class: null,
  attributes: null,

  // React props
  className: null,
  inputElement: null,
  children: null,

  // Web Component props
  custom_element: null,
  custom_method: null,
  ...renderProps
}

/**
 * The input component is an umbrella component for all inputs which share the same style as the classic `text` input field. Radio buttons and other form elements are not included here.
 */
export default class Input extends PureComponent {
  static tagName = 'dnb-input'
  static propTypes = propTypes
  static defaultProps = defaultProps
  static renderProps = renderProps

  static enableWebComponent() {
    registerElement(Input.tagName, Input, defaultProps)
  }

  static getDerivedStateFromProps(props, state) {
    const value = Input.getValue(props)
    if (
      state._listenForPropChanges &&
      value !== null &&
      value !== state.value
    ) {
      state.value = value
    }
    if (props.disabled) {
      state.inputState = 'disabled'
    }
    state._listenForPropChanges = true
    return state
  }

  static getValue(props) {
    if (props.value) return props.value
    return processChildren(props)
  }

  state = { inputState: 'virgin', value: null }

  constructor(props) {
    super(props)

    this._ref = React.createRef()
    this._id = props.id || `dnb-input-${Math.round(Math.random() * 999)}` // cause we need an id anyway

    // pass along all props we wish to have as params
    this.renderProps = pickRenderProps(props, renderProps)

    // make sure we dont trigger getDerivedStateFromProps on startup
    this.state._listenForPropChanges = true
    this.state.value = Input.getValue(props)

    // This is only to show, that we also can add a custom method to the custom element, and call it like so:
    // this._ref.current.focus() // current would be the Custom Element - we may delay it for 1 tick, to make sure we call if after it is set
    // We also have to ahve this method in here
    // setCustomElementMethod(this, 'focus', () => {
    //   setTimeout(() => {
    //     try {
    //       this._ref.current.focus()
    //     } catch (e) {
    //       //
    //     }
    //   }, 100)
    // })
  }
  onFocusHandler = event => {
    const value = event.target.value
    this.setState({
      value,
      _listenForPropChanges: false,
      inputState: 'focus'
    })
    dispatchCustomElementEvent(this, 'on_focus', { value, event })
  }
  onBlurHandler = event => {
    const value = event.target.value
    this.setState({
      value,
      _listenForPropChanges: false,
      inputState: 'dirty'
    })
    dispatchCustomElementEvent(this, 'on_blur', { value, event })
  }
  onChangeHandler = event => {
    const value = event.target.value
    this.setState({ value, _listenForPropChanges: false })
    event.preventDefault()
    dispatchCustomElementEvent(this, 'on_change', { value, event })
  }
  onKeyDownHandler = event => {
    if (event.key === 'Enter') {
      const value = event.target.value
      event.preventDefault()
      dispatchCustomElementEvent(this, 'on_submit', { value, event })
    }
  }
  render() {
    const {
      type,
      size,
      label,
      status,
      status_state,
      disabled,
      placeholder,
      description,
      align,
      input_class,
      input_typo,
      extra_information
    } = this.props

    let { autocomplete } = this.props
    if (type === 'search') {
      autocomplete = null
    }
    const id = this._id

    const classes = classnames(
      'dnb-input',
      `dnb-input--${type}`, //type_modifier
      extra_information ? 'dnb-input--has-extra-information' : null,
      size ? 'dnb-input--' + size : '',
      type === 'search' ? 'dnb-input__input--search' : null,
      align ? `dnb-input__align--${align}` : null,
      status ? `dnb-input__status--${status_state}` : null,
      this.props.class,
      this.props.className
    )

    const { inputElement: Elem, ...renderProps } = this.renderProps

    const inputParams = {
      ...renderProps,
      className: classnames('dnb-input__input', input_typo, input_class),
      autoComplete: autocomplete,
      value: this.state.value || '',
      type,
      id,
      // align,
      disabled,
      name: id,
      onChange: this.onChangeHandler,
      onKeyDown: this.onKeyDownHandler,
      onFocus: this.onFocusHandler,
      onBlur: this.onBlurHandler
    }

    // also used for code markup simulation
    validateDOMAttributes(this.props, inputParams)

    if (description) {
      inputParams['aria-describedby'] = id + '-description'
    }

    const shellParams = {
      'data-input-state': this.state.inputState,
      'data-has-content':
        String(this.state.value || '').length > 0 ? 'true' : 'false'
    }

    return (
      <span className={classes}>
        {label && (
          <FormLabel for_id={id} text={label} disabled={disabled} />
        )}

        <span className="dnb-input__shell" {...shellParams}>
          {(type === 'text' || type === 'number' || type === 'search') &&
            ((typeof Elem === 'function' ? (
              <Elem innerRef={this._ref} {...inputParams} />
            ) : null) || <input ref={this._ref} {...inputParams} />)}

          {type === 'search' && (
            <Submit
              {...this.props}
              value={inputParams.value}
              title={this.props.search_button_title}
            />
          )}

          {placeholder && (
            <span
              className={classnames(
                'dnb-input__placeholder',
                align ? `dnb-input__align--${align}` : null
              )}
            >
              {placeholder}
            </span>
          )}
        </span>

        {this.props.description && (
          <span
            className="dnb-input__description"
            id={id + '-description'}
          >
            {this.props.description}
          </span>
        )}

        {extra_information && (
          <span className="dnb-input__extra-information">
            {extra_information}
          </span>
        )}

        {status && status !== 'true' && (
          <FormStatus for_id={id} text={status} status={status_state} />
        )}
      </span>
    )
  }
}

class Submit extends PureComponent {
  static propTypes = {
    title: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    icon_size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

    // React props
    onSubmit: PropTypes.func,
    onSubmitFocus: PropTypes.func,
    onSubmitBlur: PropTypes.func,

    // Web Component props
    on_submit: PropTypes.func,
    on_submit_focus: PropTypes.func,
    on_submit_blur: PropTypes.func
  }

  static defaultProps = {
    title: null,
    icon_size: DefaultIconSize,

    // React props
    onSubmit: null,
    onSubmitFocus: null,
    onSubmitBlur: null,

    // Web Component props
    on_submit: null,
    on_submit_focus: null,
    on_submit_blur: null
  }

  state = { focusState: 'virgin' }

  onFocusHandler = event => {
    const value = this.props.value
    this.setState({
      focusState: 'focus'
    })
    dispatchCustomElementEvent(this, 'on_submit_focus', { value, event })
  }
  onBlurHandler = event => {
    const value = this.props.value
    this.setState({
      focusState: 'dirty'
    })
    dispatchCustomElementEvent(this, 'on_submit_blur', { value, event })
  }
  onSubmitHandler = event => {
    const value = this.props.value
    dispatchCustomElementEvent(this, 'on_submit', { value, event })
  }
  render() {
    const { title } = this.props
    const params = {
      type: 'submit',
      title
    }

    // also used for code markup simulation
    validateDOMAttributes(this.props, params)

    return (
      <span
        className="dnb-input__search-submit"
        data-input-state={this.state.focusState}
      >
        <Button
          className="dnb-input__search-submit__button"
          variant="secondary"
          onClick={this.onSubmitHandler}
          onFocus={this.onFocusHandler}
          onBlur={this.onBlurHandler}
          icon="search"
          {...params}
        />
      </span>
    )
  }
}
