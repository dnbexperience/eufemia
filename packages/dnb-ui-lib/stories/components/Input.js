/**
 * dnb-ui-lib Component Story
 *
 */

import React, { useState, useEffect } from 'react'
import { Wrapper, Box } from '../helpers'
import styled from '@emotion/styled'

import {
  Input,
  InputMasked,
  Button,
  FormSet,
  FormRow,
  FormLabel
} from '../../src/components'

const CustomStyle = styled.div`
  p {
    background-color: rgba(213, 30, 149, 0.25);
  }
`

export default [
  'Inputs',
  () => (
    <CustomStyle>
      <Wrapper>
        <Box>
          <Input
            label="Choose file:"
            type="file"
            accept="image/png, image/jpeg"
          />
        </Box>
        <Box>
          <CustomInput />
        </Box>
        <Box>
          <FormSet>
            <FormRow
              indent
              indent_offset="x-large"
              wrap
              label="Long label labwl Adipiscing mauris dis proin nec Condimentum egestas class blandit netus non a suscipit id urna:"
            >
              <Input label="Input A:" top="small" right="small" />
              <Input label="Input B:" top="small" right="small" />
              <Input label="Input C:" top="small" right="small" />
            </FormRow>
          </FormSet>
        </Box>
        <Box>
          <FormRow
            label="Vertical label:"
            label_direction="vertical"
            // vertical
          >
            <Input label="Input label A:" right="small" />
            <Input label="Input label B:" />
          </FormRow>
        </Box>
        <Box>
          <FormSet direction="vertical">
            <FormRow label="Legend:">
              <Input label="Vertical 1:" />
              <Input label="Vertical 2:" stretch top="small" />
            </FormRow>
          </FormSet>
        </Box>
        <Box>
          <FormSet vertical>
            <FormRow label="Legend:">
              <Input label="Vertical 1:" />
              <Input label="Vertical 2:" stretch top="small" />
            </FormRow>
          </FormSet>
        </Box>
        <Box>
          <Input
            label="Vertical label:"
            value="Stretch me ..."
            stretch
            label_direction="vertical"
          />
        </Box>
        <Box>
          Text
          <Input
            selectall
            label="Label:"
            on_change={event => {
              console.log('on_change', event)
            }}
          >
            Input ...
          </Input>
          Text
        </Box>
        <Box>
          Text
          <FormLabel>FormLabel:</FormLabel>
          <Input>Input ...</Input>
          Text
        </Box>
        <Box>
          <p className="dnb-p">
            <Input
              label="ReadOnly:"
              placeholder="Placeholder ..."
              readOnly
            />
          </p>
        </Box>
        <Box>
          <Input
            label="Search:"
            type="search"
            align="right"
            stretch
            submit_button_title="Search"
            placeholder="Search text placeholder"
          />
          <Input
            label="Search:"
            size="medium"
            type="search"
            align="right"
            stretch
            placeholder="Medium input with right aligned text"
          />
          <Input
            label="Search:"
            size="large"
            type="search"
            align="right"
            stretch
            placeholder="Large input with right aligned text"
          />
        </Box>
        <Box>
          <Input
            disabled
            status="Error"
            label="Disabled search:"
            type="search"
            submit_button_title="Search"
            placeholder="Search text placeholder"
          />
        </Box>
        <Box>
          <Input
            label="Input with status:"
            status="Message to the user"
            value="Input value with status"
          />
        </Box>
        <Box>
          <Input
            label="Input with description:"
            description="Description to the user"
            value="Input value with status"
          />
        </Box>
        <Box>
          <InputMasked
            label="Masked:"
            autocomplete="off"
            // value="1000000"
            mask={[
              '+',
              /[4]/, // have to start with 4
              /[5-7]/, // can be 5,6 or 7
              ' ',
              '/',
              ' ',
              /[49]/, // have to start with 4 or 9
              /\d/,
              /\d/,
              ' ',
              /\d/,
              /\d/,
              ' ',
              /\d/,
              /\d/,
              /\d/
            ]}
            show_mask="true"
          />
        </Box>
        <Box>
          <form
            onSubmit={event => {
              // console.log('onSubmit', event)
              event.preventDefault()
              // event.persist()
            }}
          >
            <InputUpdate />
            <Button
              text="Submit"
              type="submit"
              on_click={event => {
                console.log('on_click', event)
              }}
              onClick={event => {
                console.log('onClick', event)
              }}
            />
          </form>
        </Box>
      </Wrapper>
    </CustomStyle>
  )
]

const InputUpdate = () => {
  const [initValue, setNewValue] = useState('Input ...')

  useEffect(() => {
    const timeoutId = setTimeout(() => {
      setNewValue('')
    }, 1e3)
    return () => clearTimeout(timeoutId)
  })

  return (
    <Input
      label="Label:"
      on_change={({ value }) => {
        console.log('on_change', value)
      }}
      onChange={({ value }) => {
        console.log('onChange', value)
      }}
      on_submit={({ value }) => {
        console.log('on_submit', value)
      }}
      onSubmit={({ value }) => {
        console.log('onSubmit', value)
      }}
      value={initValue}
    />
  )
}

const CustomInput = () => {
  const [value, setValue] = useState('2019-02-15')
  return (
    <>
      <Input
        value={value}
        on_change={({ value }) => {
          console.log('on_change', value)
          setValue(value)
        }}
        on_state_update={({ value }) => {
          console.warn('on_state_update', value)
          setValue(value)
        }}
        right
      />
      <Button
        text="Reset"
        on_click={() => {
          setValue('123')
        }}
      />
    </>
  )
}
