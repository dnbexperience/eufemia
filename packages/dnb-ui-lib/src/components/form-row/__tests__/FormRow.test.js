/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  fakeProps,
  axeComponent,
  toJson,
  loadScss
} from '../../../core/jest/jestSetup'
import Component from '../FormRow'
import Input from '../../input/Input'

const props = fakeProps(require.resolve('../FormRow'), {
  optional: true
})
props.id = 'form-row'
props.direction = 'horizontal'
props.label_direction = 'horizontal'

describe('FormRow component', () => {
  const Comp = mount(<Component {...props} />)

  it('have to match snapshot', () => {
    expect(toJson(Comp)).toMatchSnapshot()
  })

  it('should have vertical direction class', () => {
    const Comp = mount(<Component {...props} direction="vertical" />)
    expect(
      Comp.find('.dnb-form-row').hasClass('dnb-form-row--vertical')
    ).toBe(true)
  })

  it('should have correct indent class', () => {
    const Comp = mount(<Component {...props} indent="large" />)
    expect(
      Comp.find('.dnb-form-row').hasClass('dnb-form-row__indent--large')
    ).toBe(true)
  })

  it('should have an isolated state on nested FormRows', () => {
    const Comp = mount(
      <Component vertical>
        <Input label="Vertical" />
        <Component vertical="false" label_direction="horizontal">
          <Input label="Horizontal" />
        </Component>
        <Input label="Vertical" />
      </Component>
    )
    expect(
      Comp.find('span.dnb-input').at(0).hasClass('dnb-input--vertical')
    ).toBe(true)
    expect(
      Comp.find('span.dnb-input').at(1).hasClass('dnb-input--horizontal')
    ).toBe(true)
    expect(
      Comp.find('span.dnb-input').at(2).hasClass('dnb-input--vertical')
    ).toBe(true)
  })

  it('should using formset and legend by default', () => {
    expect(Comp.find('fieldset').exists()).toBe(true)
    expect(Comp.find('legend').exists()).toBe(true)
  })

  it('should using formset and legend by default', () => {
    const Comp = mount(<Component {...props} no_fieldset />)
    expect(Comp.find('label').exists()).toBe(true)
    expect(Comp.find('fieldset').exists()).toBe(false)
    expect(Comp.find('legend').exists()).toBe(false)
  })

  it('should react correct on two states in row', () => {
    const Comp = mount(
      <Component {...props} disabled={false}>
        <Input />
      </Component>
    )
    Comp.setProps({ disabled: true })
    expect(Comp.find('input').is('[disabled]')).toBe(true)
  })

  it('should validate with ARIA rules', async () => {
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('FormRow scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/dnb-form-row.scss'))
    expect(scss).toMatchSnapshot()
  })
  it('have to match default theme snapshot', () => {
    const scss = loadScss(
      require.resolve('../style/themes/dnb-form-row-theme-ui.scss')
    )
    expect(scss).toMatchSnapshot()
  })
})
