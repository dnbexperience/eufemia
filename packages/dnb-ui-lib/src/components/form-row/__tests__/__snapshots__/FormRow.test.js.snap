// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRow component have to match snapshot 1`] = `
<FormRow
  class="class"
  className="className"
  custom_element={Object {}}
  custom_method={[Function]}
  direction="horizontal"
  disabled="disabled"
  global_status_id="global_status_id"
  id="form-row"
  indent="indent"
  indent_offset="indent_offset"
  label="label"
  label_class="label_class"
  label_direction="horizontal"
  label_id="label_id"
  label_sr_only="label_sr_only"
  no_fieldset="no_fieldset"
  no_label="no_label"
  render_content={[Function]}
  section_spacing="section_spacing"
  section_style="section_style"
  skipContentWrapperIfNested={true}
  vertical="vertical"
  wrap="wrap"
>
  <Fieldset
    className={null}
    useFieldset={true}
  >
    <fieldset
      className="dnb-form-row__fieldset"
    >
      <div
        className="dnb-form-row dnb-form-row--horizontal dnb-form-row--horizontal-label className class dnb-form-row__indent--indent dnb-section dnb-section--section_style dnb-section--spacing-section_spacing"
      >
        <AlignmentHelper
          className="dnb-form-row__helper"
        >
          <span
            aria-hidden={true}
            className="dnb-alignment-helper dnb-form-row__helper"
          />
        </AlignmentHelper>
        <FormLabel
          class={null}
          className="dnb-form-row__label label_class"
          direction="horizontal"
          disabled={false}
          element="legend"
          for_id={null}
          id="label_id"
          label_direction={null}
          render_content={null}
          sr_only="label_sr_only"
          text="label"
          title={null}
          vertical={null}
        >
          <Element
            className="dnb-form-label dnb-form-row__label label_class"
            id="label_id"
            is="legend"
          >
            <legend
              className="dnb-form-label dnb-form-row__label label_class"
              id="label_id"
            >
              label
            </legend>
          </Element>
        </FormLabel>
        <div
          className="dnb-form-row__content dnb-form-row__content--indent_offset"
        >
          children
        </div>
      </div>
    </fieldset>
  </Fieldset>
</FormRow>
`;

exports[`FormRow scss have to match default theme snapshot 1`] = `
"/*
 * FormRow theme
 *
 */
.dnb-form-row:not(.dnb-form-row--vertical).dnb-form-row__indent--default > .dnb-form-label, .dnb-form-row:not(.dnb-form-row--vertical).dnb-form-row__indent--medium > .dnb-form-label {
  width: 30vw;
  max-width: 20rem; }
  @media (max-width: 40em) {
    .dnb-form-row:not(.dnb-form-row--vertical).dnb-form-row__indent--default > .dnb-form-label, .dnb-form-row:not(.dnb-form-row--vertical).dnb-form-row__indent--medium > .dnb-form-label {
      width: auto; } }
"
`;

exports[`FormRow scss have to match snapshot 1`] = `
"/*
* DNB FormRow
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
/*
* DNB FormLabel
*
*/
.dnb-form-label {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-default);
  font-size: var(--font-size-small);
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-form-label *,
  .dnb-form-label ::before,
  .dnb-form-label ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-form-label ::before,
  .dnb-form-label ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
 * FormLabel component
 *
 */
.dnb-form-label {
  display: inline-block;
  width: auto;
  padding: 0;
  margin-right: 1rem;
  text-align: left;
  font-size: var(--font-size-basis);
  vertical-align: baseline;
  color: inherit; }
  .dnb-form-label--vertical {
    display: block;
    margin-right: 0;
    margin-bottom: 0.5rem; }
  .dnb-form-label[for]:not([disabled]) {
    user-select: none;
    -webkit-user-select: none;
    cursor: pointer; }
  .dnb-form-label[disabled] {
    cursor: not-allowed; }
  @media screen and (-ms-high-contrast: none) {
    .dnb-form-label {
      max-width: 60vw; } }

legend.dnb-form-label {
  display: inline-block; }

.dnb-form-row {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-default);
  font-size: var(--font-size-small);
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-form-row *,
  .dnb-form-row ::before,
  .dnb-form-row ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-form-row ::before,
  .dnb-form-row ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
 * FormRow component
 *
 */
:root {
  --form-row-column1-start: auto;
  --form-row-column2-start: 1fr;
  --form-row-row1-start: auto;
  --form-row-row2-start: auto;
  --form-bounding-height: 0.25rem; }

.dnb-form-row {
  display: flex;
  align-items: baseline; }
  .dnb-form-row__content {
    display: inline-flex;
    flex: 1;
    width: 100%; }
  .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content {
    align-items: baseline; }
    .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content .dnb-form-label {
      white-space: nowrap; }
    .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content--wrap {
      display: block; }
  @supports (display: grid) {
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] {
      display: grid;
      grid-template-columns: [column1-start] var(--form-row-column1-start) [column1-end column2-start] var(--form-row-column2-start) [column2-end];
      grid-template-rows: [row1-start] var(--form-row-row1-start) [row1-end row2-start] var(--form-row-row2-start) [row2-end]; } }
  @supports (display: grid) {
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content {
      grid-column: column2-start / column2-end;
      grid-row: row2-start / row2-end; } }
  @media (min-width: 40em) {
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--default {
      margin-top: -1.5rem; }
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--x-small {
      margin-top: calc(-0.5rem + var(--form-bounding-height)); }
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--small {
      margin-top: calc(-1rem + var(--form-bounding-height)); }
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--medium {
      margin-top: calc(-1.5rem + var(--form-bounding-height)); }
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--large {
      margin-top: calc(-2rem + var(--form-bounding-height)); }
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--m-large {
      margin-top: calc(-2.5rem + 0.625rem); }
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--x-large {
      margin-top: calc(-3rem + var(--form-bounding-height)); }
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content--xx-large {
      margin-top: calc(-3.5rem + var(--form-bounding-height)); } }
  @media screen and (-ms-high-contrast: none) {
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content {
      margin-top: 0; } }
  @media (max-width: 40em) {
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-row__content {
      margin-top: 0.5rem !important; } }
  .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-label {
    margin-right: 1rem; }
    @supports (display: grid) {
      .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-label {
        grid-column: column1-start / column1-end;
        grid-row: row1-start / row1-end; } }
    @media (min-width: 40em) {
      .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] > .dnb-form-label {
        text-align: right; } }
  @media (max-width: 40em) {
    .dnb-form-row:not(.dnb-form-row--vertical)[class*='__indent'] {
      display: block; } }
  @media (max-width: 40em) {
    .dnb-form-row:not(.dnb-form-row--vertical) {
      flex-direction: column;
      align-items: flex-start; }
      .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-label {
        margin-bottom: 0.5rem; }
      .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content {
        flex-direction: column; }
        .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content > .dnb-form-label {
          margin-top: 1rem; } }
  .dnb-form-row--vertical > .dnb-form-label {
    margin-bottom: 0.5rem; }
  .dnb-form-row--vertical > .dnb-form-row__content {
    flex-direction: column; }
  .dnb-form-row--vertical-label {
    flex-direction: column; }
  @media (max-width: 40em) {
    .dnb-form-row .dnb-form-component {
      margin-bottom: 0.5rem; } }
  .dnb-form-row > .dnb-form-status {
    margin: 0.5rem 0; }
  .dnb-form-row > .dnb-form-label--vertical {
    display: flex;
    flex-direction: column;
    align-items: flex-start; }
  @supports (display: grid) {
    .dnb-form-row > .dnb-form-label--vertical ~ .dnb-form-row__content {
      grid-column: column1-start / column1-end; } }
  .dnb-form-row__label-dummy {
    margin: 0; }
  .dnb-form-row__fieldset,
  .dnb-core-style .dnb-form-row__fieldset {
    margin: 0;
    padding: 0;
    border: none;
    width: 100%; }
  @media screen and (-ms-high-contrast: none) {
    .dnb-form-row {
      align-items: flex-start; }
      .dnb-form-row__content {
        flex: auto; } }
"
`;
